### PR TITLE
Changes to handle Hybris app override for local.properties pertaining…

### PR DESCRIPTION
**Pull request by faizalak**
… to Clustering needs, customize jvmRoute in Apache Tomcat server.xml, customize the ANT tasks executed for the Hybris App build process

### CODE REVIEW

File: base/s2i/assemble [hunk 1]
```diff
@@ -3,61 +3,106 @@
...
     exit 1
 fi
 
 # Copying Source to Hybris Server
-echo "---> Copying Hybris application source ..."
-cp -r ${HYBRIS_CONTENT_DIR}/* "${HYBRIS_HOME}"
+echo "---> Copying Hybris application source cp -r ${HYBRIS_CONTENT_DIR}/* ${HYBRIS_HOME}/hybris"
+cp -r ${HYBRIS_CONTENT_DIR}/* ${HYBRIS_HOME}/hybris
+
+#####Below section unitl the commented lines added on April 5 2021 for the Hybris Operator changes
```
Comment (miminar) [inline base/s2i/assemble]: one can always use `git blame` if he's curious about dates and authors

File: base/s2i/assemble [hunk 2]
```diff
@@ -3,61 +3,106 @@
...
+
+if [ ! -z "${SOURCE_REPO_LOCAL_PROPERTIES_OVERRIDE}" ]; then
+    echo "---> Hybris config  ${HYBRIS_HOME}/${HYBRIS_LOCAL_PROPERTIES_PATH} is replaced by ${SOURCE_REPO_LOCAL_PROPERTIES_OVERRIDE}"
+    curl -k -o ${HYBRIS_HOME}/${HYBRIS_LOCAL_PROPERTIES_PATH} ${SOURCE_REPO_LOCAL_PROPERTIES_OVERRIDE}
```
Comment (miminar) [inline base/s2i/assemble]: please wrap the arguments in double quotes, otherwise unexpected happens

File: base/s2i/assemble [hunk 3]
```diff
@@ -3,61 +3,106 @@
...
+else
+    echo "---> No Hybris ANT tasks to execute ${HYBRIS_ANT_TASK_NAMES}"
 fi
 
 echo
-echo "SAP Commerce Build Successful!"
+echo "---> SAP Commerce Build Successful!"
 
-ant ${ANT_DEPLOY_TASKS}
+#####Commented  out on April 5 2021 for the Hybris Operator changes to include customized ANT execution
```
Comment (miminar) [inline base/s2i/assemble]: no need for commented out code in a git repo

File: config/samples/hybris_v1alpha1_hybrisapp.yaml [hunk 4]
```diff
@@ -3,8 +3,11 @@ kind: HybrisApp
 metadata:
   name: hybrisapp-sample
 spec:
-  baseImageName: hybris-base
+  baseImageName: hybris-base-for-testing
   baseImageTag: v0.1
   sourceRepoRef: master
-  sourceRepoURL: https://github.com/redhat-sap/sap-commerce-app-example.git
+  sourceRepoURL: https://github.com/faizalak/sap-commerce-app-example.git
```
Comment (miminar) [inline config/samples/hybris_v1alpha1_hybrisapp.yaml:9]: Is this change intended?
If it's just for a testing, perhaps a new "testing" sample file could be created?

File: controllers/hybrisapp_controller.go [hunk 5]
```diff
@@ -259,6 +259,7 @@ func (r *HybrisAppReconciler) ensureBuildConfig(hybrisApp *hybrisv1alpha1.Hybris
 	if err != nil && errors.IsNotFound(err) {
 		// Define a new deployment
 		bc := r.createBuildConfigForHybrisApp(hybrisApp)
+		//log.Info("----> BuildConfig", "bc.Spec.Strategy.SourceStrategy.Env", bc.Spec.Strategy.SourceStrategy.Env)
```
Comment (miminar) [inline controllers/hybrisapp_controller.go:262]: maybe uncommented `log.Debug` would be better?

File: config/samples/hybris_v1alpha1_hybrisapp.yaml (hunk 4 above)
Comment (faizalak) [reply config/samples/hybris_v1alpha1_hybrisapp.yaml:9]: sap-commerce-app-example is a sample user-created CR for testing the operator. The repo now has a new property file, which is one of the changes to test. This will not impact runtime.

File: base/s2i/assemble (hunk 3 above)
Comment (faizalak) [reply base/s2i/assemble]: What do you suggest here? Should I remove all commented outlines? I understand your point. Just maintained them for posterity's sake.

File: base/s2i/assemble (hunk 2 above)
Comment (faizalak) [reply base/s2i/assemble]: Done

File: base/s2i/assemble (hunk 1 above)
Comment (faizalak) [reply base/s2i/assemble]: I removed all the commented unnecessary lines from the code

File: base/s2i/assemble (hunk 3 above)
Comment (faizalak) [reply base/s2i/assemble]: Removed the commented lines

File: config/samples/hybris_v1alpha1_hybrisapp.yaml (hunk 4 above)
Comment (miminar) [reply config/samples/hybris_v1alpha1_hybrisapp.yaml:9]: I see, thanks for the explanation.

**Comment by miminar** [comment]
Having zero understanding of SAP Commerce and minimal knowledge of the operator framework, it looks good to me.
But, doesn't it deserve a version bump?

File: api/v1alpha1/hybrisapp_types.go [hunk 6]
```diff
@@ -41,6 +41,15 @@ type HybrisAppSpec struct {
 
 	// Hybris app repository source location
 	SourceRepoContext string `json:"sourceRepoContext,omitempty"`
+
+	// Hybris app repository local.properties override location
+	SourceRepoLocalPorpertiesOverride string `json:"sourceRepoLocalPropertiesOverride,omitempty"`
```
Comment (xieshenzh) [inline api/v1alpha1/hybrisapp_types.go:46]: For optional spec, it is better to use pointer and check if it is `nil` in the controller before using it.
```suggestion
	SourceRepoLocalPorpertiesOverride *string `json:"sourceRepoLocalPropertiesOverride,omitempty"`
```

File: api/v1alpha1/hybrisapp_types.go [hunk 7]
```diff
@@ -41,6 +41,15 @@ type HybrisAppSpec struct {
...
+
+	// Hybris app Apache server.xml jvmRoute name
+	ApachejvmRouteName string `json:"apachejvmRouteName,omitempty"`
```
Comment (xieshenzh) [inline api/v1alpha1/hybrisapp_types.go:49]:
```suggestion
	ApachejvmRouteName *string `json:"apachejvmRouteName,omitempty"`
```

File: api/v1alpha1/hybrisapp_types.go [hunk 8]
```diff
@@ -41,6 +41,15 @@ type HybrisAppSpec struct {
...
+
+	// Hybris app ANT tasks
+	HybrisANTTaskNames string `json:"hybrisANTTaskNames,omitempty"`
```
Comment (xieshenzh) [inline api/v1alpha1/hybrisapp_types.go:52]:
```suggestion
	HybrisANTTaskNames *string `json:"hybrisANTTaskNames,omitempty"`
```

File: base/s2i/assemble [hunk 9]
```diff
@@ -3,55 +3,66 @@
...
-    echo "SAP Commerce Build Failed!"
-    exit 1
+# Execute multiple ANT tasks for Hybris one at a time
+if [ ! -z "${HYBRIS_ANT_TASK_NAMES}" ]; then
+    echo "---> All Hybris ANT tasks execution started for ${HYBRIS_ANT_TASK_NAMES}"
+    IFS=","
+    for v in ${HYBRIS_ANT_TASK_NAMES}
+    do
+        echo "---> Executing ant $v"
+        ant $v
```
Comment (xieshenzh) [inline base/s2i/assemble:53]: The build should fail immediately if any `ant` task fails.

File: config/samples/hybris_v1alpha1_hybrisapp.yaml [hunk 10]
```diff
@@ -4,7 +4,13 @@ metadata:
   name: hybrisapp-sample
 spec:
   baseImageName: hybris-base
-  baseImageTag: v0.1
+  baseImageTag: v0.2
   sourceRepoRef: master
-  sourceRepoURL: https://github.com/redhat-sap/sap-commerce-app-example.git
+  sourceRepoURL: https://github.com/faizalak/sap-commerce-app-example.git
```
Comment (xieshenzh) [inline config/samples/hybris_v1alpha1_hybrisapp.yaml:9]: I don't think personal information should be used in the examples.

File: config/samples/hybris_v1alpha1_hybrisapp.yaml [hunk 11]
```diff
@@ -4,7 +4,13 @@ metadata:
...
+  #sourceRepoLocalPropertiesOverride: ''
+  #apachejvmRouteName: ''
+  #hybrisANTTaskNames: ''
+  sourceRepoLocalPropertiesOverride: https://raw.githubusercontent.com/faizalak/sap-commerce-app-example/master/hybris/config/local.properties.qa1
```
Comment (xieshenzh) [inline config/samples/hybris_v1alpha1_hybrisapp.yaml:13]: same here

File: base/s2i/assemble [hunk 12]
```diff
@@ -3,55 +3,66 @@
...
+    done
+    echo "---> All Hybris ANT tasks execution completed for ${HYBRIS_ANT_TASK_NAMES}"
+else
+    echo "---> No Hybris ANT tasks to execute ${HYBRIS_ANT_TASK_NAMES}"
+    echo "---> Executing >ant with default task clean all"
+    ant clean all
```
Comment (xieshenzh) [inline base/s2i/assemble:59]: same here.

**Review by tejones — APPROVED** [review]
LGTM

File: controllers/hybrisapp_controller.go [hunk 13]
```diff
@@ -344,9 +345,11 @@ func (r *HybrisAppReconciler) createBuildConfigForHybrisApp(hybrisApp *hybrisv1a
 				Strategy: buildv1.BuildStrategy{
 					Type: buildv1.SourceBuildStrategyType,
 					SourceStrategy: &buildv1.SourceBuildStrategy{
-						From: corev1.ObjectReference{
-							Kind: "ImageStreamTag",
-							Name: strings.Join([]string{hybrisApp.Spec.BaseImageName, hybrisApp.Spec.BaseImageTag}, ":"),
+						From: corev1.ObjectReference{Kind: "ImageStreamTag", Name: strings.Join([]string{hybrisApp.Spec.BaseImageName, hybrisApp.Spec.BaseImageTag}, ":")},
+						Env: []corev1.EnvVar{
```
Comment (xieshenzh) [inline controllers/hybrisapp_controller.go:349]: You only covered the scenario when a new buildconfig is created. 
In the function `ensureBuildConfig`, you should also check if user updates the settings of a deployed app. If settings are updated, you should update the buildconfig and make sure a new build starts to rebuild the app image, and also the deploymentconfig redeploys the app with the new image. 
This is the scenario to test for changing settings.
By the way, the app image will be rebuilt every time the user changes some local properties, it is an issue of the approach.

**Review by xieshenzh — CHANGES_REQUESTED** [review]
Thanks @faizalak . Please also update the docs to provide some explanation on the new features.
An operator is not just an installer. It is also responsible for configuring and managing the application. That's why there should be test cases for configuring and managing the application.

File: api/v1alpha1/hybrisapp_types.go (hunk 6 above)
Comment (faizalak) [reply api/v1alpha1/hybrisapp_types.go:46]: I do not see these types being different from the existing ones, that the existing design needs to be altered. There is a check happening in assemble script, and the status gets ignored if the values are empty.

File: api/v1alpha1/hybrisapp_types.go (hunk 7 above)
Comment (faizalak) [reply api/v1alpha1/hybrisapp_types.go:49]: I do not see these types being different from the existing ones, that the existing design needs to be altered. There is a check happening in assemble script, and the status gets ignored if the values are empty.

File: api/v1alpha1/hybrisapp_types.go (hunk 8 above)
Comment (faizalak) [reply api/v1alpha1/hybrisapp_types.go:52]: I do not see these types being different from the existing ones, that the existing design needs to be altered. There is a check happening in assemble script, and the status gets ignored if the values are empty.

File: base/s2i/assemble (hunk 9 above)
Comment (faizalak) [reply base/s2i/assemble:53]: I am making a change here to break on ANT task failure. I agree with you here

File: config/samples/hybris_v1alpha1_hybrisapp.yaml (hunk 10 above)
Comment (faizalak) [reply config/samples/hybris_v1alpha1_hybrisapp.yaml:9]: I will update this

File: base/s2i/assemble (hunk 12 above)
Comment (faizalak) [reply base/s2i/assemble:59]: This is what was executed before. So why change it now?

File: controllers/hybrisapp_controller.go (hunk 13 above)
Comment (faizalak) [reply controllers/hybrisapp_controller.go:349]: This design is nothing new from before. I do not want to change the existing behavior. We can handle an architectural change as this becomes a product

**Comment by faizalak** [comment]
I will update the documentation in due course of time

File: api/v1alpha1/hybrisapp_types.go (hunk 6 above)
Comment (xieshenzh) [reply api/v1alpha1/hybrisapp_types.go:46]: The existing ones are not optional. If changing them to use pointers, the pointers will never be `nil`.

File: base/s2i/assemble (hunk 12 above)
Comment (xieshenzh) [reply base/s2i/assemble:59]: You deleted the lines that check the outcome of this.

File: controllers/hybrisapp_controller.go (hunk 13 above)
Comment (xieshenzh) [reply controllers/hybrisapp_controller.go:349]: This is not architectural change. This is how every operator works.
You can take a look at `ensureBuildConfig`, there is code to update the buildconfig when specs of the app changes.

File: api/v1alpha1/hybrisapp_types.go (hunk 6 above)
Comment (faizalak) [reply api/v1alpha1/hybrisapp_types.go:46]: I see all the existing ones being optional (omitempty) and being capable of being set with a nill value. Why do you claim they cannot be nil? I don't want to invalidate my testing now, by changing it to the pointer at this time. I will do that if future changes come up

	BaseImageName string `json:"baseImageName,omitempty"`

	// Hybris base image tag
	// +kubebuilder:validation:Required
	BaseImageTag string `json:"baseImageTag,omitempty"`

	// Hybris app source repository URL
	// +kubebuilder:validation:Required
	SourceRepoURL string `json:"sourceRepoURL,omitempty"`

	// Hybris app source repository reference
	// +kubebuilder:validation:Required
	SourceRepoRef string `json:"sourceRepoRef,omitempty"`

	// Hybris app repository source location
	SourceRepoContext string `json:"sourceRepoContext,omitempty"`

	// Hybris app repository local.properties override location
	SourceRepoLocalPorpertiesOverride string `json:"sourceRepoLocalPropertiesOverride,omitempty"`

	// Hybris app Apache server.xml jvmRoute name
	ApachejvmRouteName string `json:"apachejvmRouteName,omitempty"`

	// Hybris app ANT tasks
	HybrisANTTaskNames string `json:"hybrisANTTaskNames,omitempty"`

**Comment by xieshenzh** [comment]
@faizalak Since your code is already merged, feel free to take or ignore my comments.

File: api/v1alpha1/hybrisapp_types.go (hunk 6 above)
Comment (xieshenzh) [reply api/v1alpha1/hybrisapp_types.go:46]: Because of this annotation `kubebuilder:validation:Required`.

File: controllers/hybrisapp_controller.go (hunk 13 above)
Comment (faizalak) [reply controllers/hybrisapp_controller.go:349]: When you are free get on a call with me. Show me exactly what you are trying to refer to here? I do not see any changes in the update of the BC design. The trigger for them is as before

Comment (xieshenzh) [reply controllers/hybrisapp_controller.go:349]: It is here in `ensureBuildConfig`: https://github.com/faizalak/sap-commerce-operator/blob/0c23f99509f1860540355627c8ddc7cae713836a/controllers/hybrisapp_controller.go#L291

Comment (faizalak) [reply controllers/hybrisapp_controller.go:349]: I understand your point now. I think it is alright to proceed since I am not expecting the HybriApp to be updated later to the initial creation from the operator. These new properties need not trigger an update. Just creation should suffice for now.
I will hold this change for a future date if the customer comes back asking for anything further. For now, I need the priority to have the approach tested to see if it fits the customer

File: api/v1alpha1/hybrisapp_types.go (hunk 6 above)
Comment (faizalak) [reply api/v1alpha1/hybrisapp_types.go:46]: I don't see the code-breaking if these new properties are omitted. I tested now with a HybriApp CR having only the below properties without the new ones included. It behaved just fine.

HybrisApp CR----->
  spec:
    baseImageName: hybris-base
    baseImageTag: v0.2
    sourceRepoRef: master
    sourceRepoURL: https://github.com/faizalak/sap-commerce-app-example.git
    
Build---->
STEP 7: USER 1001
STEP 8: RUN /usr/local/s2i/assemble
---> Copying Hybris application source cp -r /tmp/src/hybris/* /opt/hybris
---> Hybris config /opt/hybris/config/local.properties present in the source repo will be used as is
---> Apache config /opt/hybris/config/tomcat/conf/server.xml present in the source repo will be used as is
---> Building SAP Commerce ...
---> No Hybris ANT tasks to execute
---> Executing >ant with default task clean all
Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=400m; support was removed in 8.0
Buildfile: /opt/hybris/bin/platform/build.xml
